src/libraries/System.Net.Security/src/System/Net/Security/SslStreamCertificateContext.Windows.cs

```diff
@@ -7,10 +7,103 @@ namespace System.Net.Security
 {
     public partial class SslStreamCertificateContext
     {
+        // No leaf, include root.
+        private const int _trimChain = 1;
+
         internal static SslStreamCertificateContext Create(X509Certificate2 target)
         {
             // On Windows we do not need to build chain unless we are asked for it.
             return new SslStreamCertificateContext(target, Array.Empty<X509Certificate2>());
         }
+
+
+        private SslStreamCertificateContext(X509Certificate2 target, X509Certificate2[] intermediates)
+        {
+            if (intermediates.Length > 0)
+            {
+                using (X509Chain chain = new X509Chain())
+                {
+                    chain.ChainPolicy.VerificationFlags = X509VerificationFlags.AllFlags;
+                    chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
+                    chain.ChainPolicy.DisableCertificateDownloads = true;
+                    bool osCanBuildChain = chain.Build(target);
+
+                    int count = 0;
+                    foreach (X509ChainStatus status in chain.ChainStatus)
+                    {
+                        if (status.Status.HasFlag(X509ChainStatusFlags.PartialChain) || status.Status.HasFlag(X509ChainStatusFlags.NotSignatureValid))
+                        {
+                            osCanBuildChain = false;
+                            break;
+                        }
+
+                        count++;
+                    }
+
+                    // OS failed to build the chain but we have at least some intermediates.
+                    // We will try to add them to "Intermediate Certification Authorities" store.
```

Comment:
Do you really want to add all intermediates?  Or just the ones that the OS ends up using in a chain?

e.g. Add all of intermediates to chain.ChainPolicy.ExtraStore and build again; then add anything relevant (preferably using a diffing algorithm so that it doesn't potentially reset custom store property overrides).

Using a second chain object (you could copy the policy object over) would make the diff easy, just walk each position and compare that cert.RawData values SequenceEqual.

Ideally, after each add you'd rerun the offline/no-extra build to see if you can avoid invalidating existing higher-scoped entries.

Reply:
(While a diffing algorithm and a while loop sound potentially expensive, most chains are three items long, so there aren't many iterations)

Reply:
I was considering to add some more complicated logic. But as you said if most chains are 3 long, it does not matter that much. I assumed that adding intermediate once to store is cheap enough so I decided not to care.

Reply:
There's one thing to check:

* Add an intermediate to the store
* Open the intermediates store in MMC (e.g. add snap-in, certificates, local computer, then pick Intermediate Certificate Authorities / Certificates in the nav tree)
* Pick that same certificate, right click, Properties
* Pick "Enable only the following purposes"
* Toggle some checkmarks
* Use the API as you have it
* Right click on the Certificates entry in the nav bar and refresh
* Reopen the properties dialog.

If the property resets, we need to do complex stuff here.  If it doesn't, we're good.  (I think we do "merge properties", which should (generally) be safe, but it's good to test that before we run the risk of altering machine state)

Reply:
that assumes same CA set between runs, right? (right now we generate unique set for each test run and I had situation where it would build up before adding cleanup to the tests)

Reply:
It assumes that we ever get something in the `intermediates` collection that's already in the store.  E.g. if what we need is a 4 hop chain:

* root -> intermed1 -> intermed2 -> end-entity

And intermed1 is already in the store, but intermed2 wasn't; we'll end up calling store.Add(intermed1) which will cause the merge logic to apply.  Since it changes system persisted state we need to be careful to ensure it's only additive, not destructive.

Reply:
here is what I did:
- modified the test helper to save/load certificates from pfx and I disabled the cleanup
- I run the test and I check both CA.
- I removed root CA, run tests and verified that it triggers the adding logic
- I removed root again and I modified properties on intermediate and I also added friendly name
- run test again. Root CA was added again and intermediate preserved with my modifications

so the conclusion is that adding existing certificate is _not_ destructive.